Split 'communicate with javascript' page into multiple pages

feihong opened the pull request:
Addresses https://github.com/melange-re/melange-re.github.io/issues/198. I basically just split the page up into multiple pages as described in the issue and then fixed all the broken links.

I didn't make a short tutorial for bindings in this PR, but I plan to do it in another one.

feihong commented on docs/how-to-guides.md:
```diff
@@ -537,14 +537,15 @@ v3, but node extensions (`%bs.*`) are not.
 The `@bs.val` attribute is no longer necessary, and can be removed from
 `external` definitions. See more information in the ["Using global functions or
-values"](./communicate-with-javascript.md#using-global-functions-or-values)
+values"](/working-with-js-objects-and-values.md#using-global-functions-or-values)
```
@jchavarri what's the convention for links? (is there one?) 

Should I:

- Add `.md` at the end of page names?
- Add `/`, `./`, or nothing at the beginning of links?

jchavarri replied:
I don't think there's one. Maybe we could keep the `.md` at the end (i am not sure if vitepress works without that), and relative paths `./`.

feihong replied:
Made all the links between pages use `./` and `.md` in https://github.com/melange-re/melange-re.github.io/pull/208/commits/c07670105875f978ed75849d2b36da47130343bc

feihong commented on docs/advanced-js-interop.md:
```diff
@@ -0,0 +1,659 @@
+# Advanced JavaScript interoperability
```
@jchavarri Should I create more structure by making a `communicate-with-javascript` directory and putting this and the other bindings topics pages into it? Or is flat page hierarchy OK?

jchavarri replied:
Not sure. I guess it makes sense to reflect the website public hierarchy on the file system as well (i.e. create a folder). @davesnx wdyt?

davesnx commented on docs/communicate-with-javascript.md:
```diff
@@ -1,3822 +1,39 @@
 
-Note that this attribute requires the return type of the binding to be an
-`option` type as well.
+Because the topic of bindings is quite big, we've broken it up into multiple
```
Why mention the breaking?

feihong replied:
Oops, that's a mistake. Fixed in https://github.com/melange-re/melange-re.github.io/pull/208/commits/96b04514c2b8c5c7d6a3a4e6188b05b6c16242e8

davesnx commented on docs/language-concepts.md:
```diff
@@ -0,0 +1,450 @@
+# Language concepts
```
I would call this Externals or FFI

feihong replied:
But the content of this page is a bit broader than just external and FFI, such as pipe operators

davesnx commented on docs/.vitepress/config.mts:
```diff
@@ -118,17 +118,25 @@ export default defineConfig({
+      {
+        text: 'Communicate with JavaScript',
+        items: [
+          { text: "Overview", link: "/communicate-with-javascript" },
```
Does overview appear on the sidebar?

davesnx reviewed (commented):
Do you thing each childn page should like to the next one?

feihong reviewed (commented):
> Do you thing each childn page should like to the next one?

@davesnx The way it's set up right now, it already does this

jchavarri reviewed (commented):
One way in which this change breaks my workflow is that sometimes I would copy the whole bindings markdown and paste it into chatgpt to start a session to write bindings. After this change, I can't do that so easily.

Is there a way around it that doesn't involve going through 4 or 5 md files?

feihong commented:
> One way in which this change breaks my workflow is that sometimes I would copy the whole bindings markdown and paste it into chatgpt to start a session to write bindings. After this change, I can't do that so easily.

@jchavarri Would it help if we add a build step that concatenates all the .md files into a single huge one and then deploy it to a subdomain, say, llm.melange.re?

jchavarri commented:
we discussed offline that my use case with chatgpt doesn't really work that well (i had different memory 😓 ) so we can ignore that and move on 👍

davesnx reviewed (approved):
This needs to get merged asap or you might lose a lot of time with conflicts

feihong commented:
@jchavarri I added `extract-code-block` rules for the new markdown files I added in https://github.com/melange-re/melange-re.github.io/pull/208/commits/2149825041fe173fbeca8d3d0e66f462f37e8020

But now I'm stuck: when I run `make check-extracted-code-blocks`, I expect to see some formatting errors, but I don't, even though the code snippets haven't yet been updated for reason 3.14. What am I missing?

jchavarri commented:
> when I run `make check-extracted-code-blocks`, I expect to see some formatting errors

this command doesn't check formatting. it checks that the code blocks can be built properly, or prints errors / warnings otherwise.

jchavarri commented:
it seems a lot of the snippets don't compile find (grep for e.g. "Unbound"). i think this could be fixed in another PR.

feihong commented:
> it seems a lot of the snippets don't compile find (grep for e.g. "Unbound"). i think this could be fixed in another PR.

Opened a new PR here: https://github.com/melange-re/melange-re.github.io/pull/211

yawaramin commented on docs/bindings-cookbook.md:
Hi, this file is basically a copy of my project 'BuckleScript Bindings Cookbook' with added OCaml translations of the examples: https://github.com/yawaramin/bucklescript-bindings-cookbook

That project is licensed under MIT. Please include the copyright and license notice as instructed in the repo: https://github.com/yawaramin/bucklescript-bindings-cookbook/blob/master/LICENSE

cc @anmonteiro

jchavarri replied:
Added in https://github.com/melange-re/melange-re.github.io/commit/0696ac3f99143eb7ab331bcd3e5ee82de085079b. Thanks for your work on the bucklescript/rescript cookbook.
